src/components/Output/custom-els/RangeInput/index.ts

```diff
@@ -89,60 +46,66 @@ class RangeInputElement extends HTMLElement {
   }
 
   attributeChangedCallback(name: string, oldValue: string, newValue: string) {
-    if (this._fireChange === false) return;
-    if (name === 'label-precision') {
-      this.labelPrecision = newValue;
-    } else if (name === 'value') {
-      this.value = newValue;
-    } else {
-      for (const propertyName in OBSERVED_ATTRIBUTES) {
-        if (OBSERVED_ATTRIBUTES[propertyName] === name) {
-          if (this[propertyName as REFLECTED_PROPERTIES] !== newValue) {
-            this._input.setAttribute(name, newValue);
-            this._update();
-          }
-          return;
-        }
-      }
-    }
+    if (this._ignoreChange) return;
+    this._input.setAttribute(name, newValue);
+    this._reflectAttributes();
+    this._update();
   }
 
   @bind
-  private _handleEvent(event: Event) {
-    this._update();
+  private _retargetEvent(event: Event) {
     event.stopImmediatePropagation();
     const retargetted = new Event(event.type, event);
     this.dispatchEvent(retargetted);
   }
 
   @bind
   private _update() {
-    this._reflectAttributes();
     const value = parseFloat(this.value || '0');
-    const min = parseFloat(this._input.min || '0');
-    const max = parseFloat(this._input.max || '100');
+    const min = parseFloat(this.min || '0');
+    const max = parseFloat(this.max || '100');
+    const labelPrecision = parseFloat(this.labelPrecision || '0');
     const percent = 100 * (value - min) / (max - min);
+    const displayValue = labelPrecision ? value.toPrecision(labelPrecision) :
+      Math.round(value).toString();
+
     this.style.setProperty('--value-percent', percent + '%');
-    const labelPrecision = this.labelPrecision;
-    let displayValue = '' + value;
-    if (labelPrecision === 0) {
-      displayValue = '' + Math.round(parseFloat('' + displayValue));
-    } else if (labelPrecision != null) {
-      displayValue = parseFloat('' + displayValue).toPrecision(this.labelPrecision as number | 0);
-    }
     this._valueDisplay.textContent = displayValue;
   }
 
   private _reflectAttributes() {
+    this._ignoreChange = true;
     for (const attributeName in REFLECTED_ATTRIBUTES) {
       const attributeValue = this._input.getAttribute(attributeName);
-      if (this.getAttribute(attributeName) !== attributeValue) {
-        this.setAttribute(attributeName, attributeValue || '');
-      }
+      if (attributeValue === null) continue;
+      this.setAttribute(attributeName, attributeValue);
     }
+    this._ignoreChange = false;
   }
 }
 
+interface RangeInputElement {
+  name: string;
+  min: string;
+  max: string;
+  step: string;
+  value: string;
+  disabled: boolean;
+}
+
+for (const prop of REFLECTED_PROPERTIES) {
+  Object.defineProperty(RangeInputElement.prototype, prop, {
+    get() {
+      return this._input[prop];
+    },
+    set(val) {
+      this._input[prop] = val;
+      this._reflectAttributes();
+      this._update();
+    },
+  });
+}
```

Comment:
There's some type mess here, but it feels pretty well contained, and doesn't lose types from the outside (like cleanMerge does)

Reply:
I'd take type mess over the verbosity I ended up having to avoid it.